tests/purs/passing/DerivingTraversable.purs

```diff
@@ -0,0 +1,53 @@
+module Main where
+
+import Prelude
+
+import Effect.Console (log)
+import Data.Eq (class Eq1)
+import Data.Foldable (class Foldable)
+import Data.Traversable (class Traversable, traverse, sequence)
+import Test.Assert
+
+data M f a
+  = M0
+  | M1 a (Array a)
+  | M2 Int
+  | M3 (f a)
+
+derive instance (Eq1 f, Eq a) => Eq (M f a)
+derive instance Functor f => Functor (M f)
+derive instance Foldable f => Foldable (M f)
+derive instance Traversable f => Traversable (M f)
```

Comment:
I think if we derive `Traversable`, should be automatically derived `Functor` and `Foldable`

Reply:
@rhendric just reaction and no some explanation?

Reply:
I'd disagree too, because this would be the only place that deriving one class magically makes others derived too - we don't automatically derive `Eq` from `Ord` for instance.

It could be argued that we should do it there too, but I think being fully explicit is better. It might be that there's some reason to write one of these instances by hand. It's true that we could do something to manage that, and only derive if there's no explicit instances, etc. but I think it's just better to keep it simple and only do explicitly what is asked for.

Reply:
I'd rather be explicit.

Reply:
I'm not really aware of any language that does deriving implicitly as well

Reply:
@xgrommx Did you mean that they will be automatically derived by the compiler, or did you mean that there should be a check that they also were derived (instead of being hand-written)? If the latter, I think that is a good idea.

This is a related issue that I opened a while ago (and ran into again recently!): https://github.com/purescript/purescript/issues/3168

I think we planned on elevating that particular case to an error, but never did so at a breaking change.

I think it might be worth just adding a global check that the derived instances are compatible. Especially if we support `DerivingVia` later, this could get tricky. I'll consider moving that forward later if we agree that's a good idea.

Reply:
Oh, I don't know about that. It's a real safety problem if superclasses of newtype-derived instances aren't derived because of how the compiler handles newtypes. I don't think there's an equivalent problem for `Functor` and friends; as long as the relevant laws are satisfied, nothing will break with a manually-written `Functor` and a derived `Traversable`.

Reply:
Hm that's true, But a manually-written _`Foldable`_ and derived `Traversable` may disagree on the order of the fields, which would be a problem.

Reply:
In that case the pair of instances would violate the
```purs
foldMap f = unwrap <<< traverse (Const <<< f)
```
law. I still see that as a case where if you write an instance, it's on you to ensure that it's lawful.

(I guess the twist is that this is thought of as a `Traversable` law, and the user is writing the `Foldable` instance. Morally I don't think this changes anything—to my mind, `Foldable` and `Traversable` share joint ownership of that law—but I suppose it is unusual to be worried about implementing an instance so that a subclass's law can be satisfied.)

Reply:
Is there any reason why we can't add the foldable/traversable law to both type classes and call it a day?